### PR TITLE
Implement RRT_star

### PR DESCRIPTION
**Pull request by pranavjad**
Implemented RRT_star and add python bindings. Also added "ao_benchmark.py" script to generate iterations vs solution length graph.

### CODE REVIEW

File: src/impl/vamp/planning/rrt_star.hh [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+                if (it != children[parent_node].end())
+                {
+                    children[parent_node].erase(it);
+                }
+            };
+            const std::function<void(std::size_t)> update_children =
+                [&cost, &children, &buffer_index, &update_children](std::size_t idx)
+            {
+                Configuration cur(buffer_index(idx));
+                for (auto &child : children[idx])
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh]: can be `const auto &`

File: src/impl/vamp/planning/rrt_star.hh [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+            const auto remove_edge = [&parent, &children](std::size_t parent_node, std::size_t child_node)
+            {
+                parent[child_node] = -1;
+                auto it = std::find(children[parent_node].begin(), children[parent_node].end(), child_node);
+                if (it != children[parent_node].end())
+                {
+                    children[parent_node].erase(it);
+                }
+            };
+            const std::function<void(std::size_t)> update_children =
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh:79]: why explicit `std::function` definition and not `auto`?

File: src/impl/vamp/planning/rrt_star.hh [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+                children[parent_node].emplace_back(child_node);
+            };
+            const auto remove_edge = [&parent, &children](std::size_t parent_node, std::size_t child_node)
+            {
+                parent[child_node] = -1;
+                auto it = std::find(children[parent_node].begin(), children[parent_node].end(), child_node);
+                if (it != children[parent_node].end())
+                {
+                    children[parent_node].erase(it);
+                }
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh:76]: No need to change right now, but worried about performance with `erase` - that might cause a lot of memory copies if children are early in the list.

File: src/impl/vamp/planning/rrt_star.hh [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+                vamp::utils::vector_alloc<float, FloatVectorAlignment, FloatVectorWidth>(
+                    settings.max_samples * Configuration::num_scalars_rounded),
+                &free);
+
+            const auto buffer_index = [&buffer](std::size_t index) -> float *
+            { return buffer.get() + index * Configuration::num_scalars_rounded; };
+
+            std::vector<std::size_t> parent(settings.max_samples);
+            std::vector<std::vector<std::size_t>> children(settings.max_samples, std::vector<std::size_t>());
+            std::vector<float> cost(settings.max_samples, -1.0);
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh]: Does this need to be initialized?

File: src/impl/vamp/planning/rrt_star.hh [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+                    float *new_configuration_ptr = buffer_index(free_index);
+                    auto new_configuration = nearest_configuration + extension_vector;
+                    new_configuration.to_array(new_configuration_ptr);
+                    NNNode<dimension> new_node({free_index, {new_configuration_ptr}});
+
+                    // sample points in graph within radius r of new config
+                    const std::size_t card =
+                        tree.size() + 1;  // plus 1 to account for the new configuration not inserted yet
+                    const float rrt_radius = std::fmin(
+                        settings.rewire_factor * gamma_rrt * pow((log(card) / card), dim_recip),
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh]: std::pow, std::log for consistency

File: src/impl/vamp/planning/rrt_star.hh [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+
+                    // initialize variables to keep track of the min cost neighbor and the cost of that
+                    // neighbor
+                    auto min_neighbor = nearest_node;
+                    float min_cost = cost[nearest_node.index] + nearest_distance;
+
+                    // loop through near neighbors, find the one that gives us min cost from root
+                    std::vector<bool> collision_free(near.size());
+                    std::size_t idx = 0;
+                    for (auto &[node, distance] : near)
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh]: `const auto &`?

File: src/impl/vamp/planning/rrt_star.hh [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+                        idx++;
+                    }
+
+                    // add an edge from min cost neighbor to new config
+                    tree.insert(new_node);
+                    add_edge(min_neighbor.index, free_index);
+                    cost[free_index] = min_cost;
+
+                    // rewire with our newly added node to if we can use it to get a shorter path to any of
+                    // it's neighbors
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh]: its

File: src/impl/vamp/planning/rrt_star.hh (hunk 6 above)
Comment (zkingston) [reply src/impl/vamp/planning/rrt_star.hh]: Also, I would just iterate through with `idx` as a loop variable (auto i = 0U; ...), and access near with:
```
const auto &[node, distance] = near[idx];
```

File: src/impl/vamp/planning/rrt_star.hh [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+                            {
+                                goal_motions.emplace_back(free_index - 1, goal);
+                                check_new_best = true;
+                            }
+                        }
+                    }
+                    if (not settings.force_max_iters and goal_reached)
+                    {
+                        break;
+                    }
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh:248]: Add some spaces after each block, there are a few times this happens.

File: src/impl/vamp/planning/rrt_star.hh [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+                    if (settings.force_max_iters and check_new_best)
+                    {
+                        // loop through goal nodes to find the best solution
+                        for (auto &[end_node_idx, goal] : goal_motions)
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh]: const

Comment (zkingston) [reply src/impl/vamp/planning/rrt_star.hh]: Why are you tracking multiple goal motions? You should just be able to keep the best one.

**Review by zkingston — CHANGES_REQUESTED** [review]
Few more nits

File: src/impl/vamp/planning/rrt_star.hh (hunk 2 above)
Comment (pranavjad) [reply src/impl/vamp/planning/rrt_star.hh:79]: It is recursive so I get a compile error when using `auto`.

File: src/impl/vamp/planning/rrt_star.hh (hunk 4 above)
Comment (pranavjad) [reply src/impl/vamp/planning/rrt_star.hh]: No, it was left over from debugging.

File: scripts/ao_benchmark.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    df = pd.DataFrame.from_dict(all_results)
+
+    # Convert to microseconds
+    df["planning_time"] = df["planning_time"].dt.microseconds
+    df["simplification_time"] = df["simplification_time"].dt.microseconds
+
+    print(df)
+
+    # Create the plot
+    if (plot):
```
Comment (zkingston) [inline scripts/ao_benchmark.py]: Don't need ()

File: src/impl/vamp/planning/rrt_star.hh [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+                            if (cur_cost < best_path_cost)
+                            {
+                                best_path_cost = cur_cost;
+                                best_goal_motion = {end_node_idx, goal};
+                            }
+                        }
+                    }
+                }
+            }
+            if (not settings.force_max_iters)
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh]: Space in betwen

File: src/impl/vamp/planning/rrt_star.hh [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+            {
+                result.nanoseconds = vamp::utils::get_elapsed_nanoseconds(start_time);
+                result.iterations = iter;
+                result.size.emplace_back(tree.size());
+                return result;
+            }
+            if (goal_motions.size() > 0)
+            {
+                result.cost = best_path_cost;
+                auto &[best_end_node_idx, best_goal] = best_goal_motion;
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh]: const?

File: src/impl/vamp/planning/rrt_star.hh [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+                    {
+                        break;
+                    }
+
+                    if (settings.force_max_iters and check_new_best)
+                    {
+                        // loop through goal nodes to find the best solution
+                        for (const auto &[end_node_idx, goal] : goal_motions)
+                        {
+                            float cur_cost = cost[end_node_idx] + goal.distance(buffer_index(end_node_idx));
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh]: const

File: src/impl/vamp/planning/rrt_star.hh [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+                                result.path.emplace_back(goal);
+                                result.path.emplace_back(new_configuration);
+                                result.cost = cost[current] + goal.distance(buffer_index(current));
+                                while (parent[current] != current)  // build path
+                                {
+                                    auto parent_node = parent[current];
+                                    result.path.emplace_back(buffer_index(parent_node));
+                                    current = parent_node;
+                                }
+                                std::reverse(result.path.begin(), result.path.end());
```
Comment (zkingston) [inline src/impl/vamp/planning/rrt_star.hh]: Possibly extract to a submethod? Used twice in the code.

**Review by zkingston — APPROVED** [review]
LGTM, small nits and you are good to open the pull on the main repo.

**Pull request merged**